posthog/caching/calculate_results.py

```diff
@@ -49,40 +49,62 @@ def calculate_cache_key(target: Union[DashboardTile, Insight]) -> Optional[str]:
     return generate_insight_cache_key(insight, dashboard)
 
 
-def get_cache_type(filter: FilterType) -> CacheType:
-    if filter.insight == INSIGHT_FUNNELS:
-        return CacheType.FUNNEL
-    elif filter.insight == INSIGHT_PATHS:
-        return CacheType.PATHS
-    elif filter.insight == INSIGHT_RETENTION:
-        return CacheType.RETENTION
+def get_cache_type(filter: Optional[FilterType], query: Optional[Dict]) -> CacheType:
+    if not filter and not query:
```

Comment:
Can you pass both?

Reply:
Potentially yes, if we have a migration or an interim period we might have insights that have both a filter and a query set. 

Wherever there's a query we should prefer to use that if there are both

Reply:
Would it make sense to:

1. Give query a stricter typing than just `Dict`
2. Pass `Union[FilterType, Query]` to functions

This hides the complexity and don't make me think.

Reply:
Query does have that typing but I'd hidden it so that the cache didn't have to care 🙈 🤣

Reply:
Once we've replaced filters the method would be

query = get_query(data=insight.dashboard_query(dashboard), team=team)
    cache_key = generate_insight_cache_key(insight, dashboard)
    cache_type = get_cache_type(query)

    tag_queries(
        team_id=team.pk,
        insight_id=insight.pk,
        cache_type=cache_type,
        cache_key=cache_key,
    )

    return cache_key, cache_type, process_query(team, query, True)
Inside process_query the Dict becomes a type from schema.py but from the point of view of the cache it loads some JSON from an insight and stores some JSON in the cache

Reply:
How about https://github.com/PostHog/posthog/pull/14619/commits/eb1c767d9db183bd0089ff703b90409c18618159 to separate them more and make the direction a little more explicit?